### PR TITLE
fix: Fix some uninitialised memory errors found by valgrind.

### PR DESCRIPTION
**Pull request by iphydf**
Also added a valgrind build to run it on every pull request. I've had to
disable a few tests because valgrind makes those run infinitely slowly,
consistently timing them out.

---
This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1877)

### CODE REVIEW

**Review by sudden6 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1877)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @iphydf)

___
*[toxcore/crypto_core.c, line 155 at r2](https://reviewable.io/reviews/toktok/c-toxcore/1877#-MtJ4Oe_7w0To8kJKe6F:-MtJ4Oe_7w0To8kJKe6G:b-wwrany) ([raw file](https://github.com/toktok/c-toxcore/blob/2b399eba46ed6d405a8aade73768c235149bdf5d/toxcore/crypto_core.c#L155)):*
> ```C
> #ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
>     // Avoid false alarms of uninitialized memory.
>     memset(encrypted, 0x99, length + crypto_box_MACBYTES);
> ```

why did you change the order and size here? if the order is causing errors it might be a bug

**Review by JFreegman — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1877)* status: :shipit: complete! 1 change requests, 1 of 1 approvals obtained (waiting on @iphydf)

___
*[toxcore/crypto_core.c, line 58 at r2](https://reviewable.io/reviews/toktok/c-toxcore/1877#-MtJ9jOIA5qSVz-vEqrA:-MtJ9jOIA5qSVz-vEqrB:b-hubg8c) ([raw file](https://github.com/toktok/c-toxcore/blob/2b399eba46ed6d405a8aade73768c235149bdf5d/toxcore/crypto_core.c#L58)):*
> ```C
> 
> #if !defined(FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION)
> static uint8_t *crypto_malloc(size_t bytes)
> ```

Why does this function exist? It seems like it would be a good opportunity to use memlock here (for another PR maybe)
___
*[toxcore/crypto_core.c, line 200 at r2](https://reviewable.io/reviews/toktok/c-toxcore/1877#-MtJ8vsxCBHvWx_HhYlE:-MtJ8vsxCBHvWx_HhYlF:bmjq9af) ([raw file](https://github.com/toktok/c-toxcore/blob/2b399eba46ed6d405a8aade73768c235149bdf5d/toxcore/crypto_core.c#L200)):*
> ```C
> 
> #ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
>     memcpy(plain, encrypted, length - crypto_box_MACBYTES);  // Don't encrypt anything
> ```

This isn't really part of the PR but we need to make sure `length` is > `crypto_box_MACBYTES` before this memcpy.

File: toxcore/crypto_core.c [hunk 1]
```diff
@@ -57,7 +57,13 @@ static_assert(CRYPTO_PUBLIC_KEY_SIZE == 32,
...
 static uint8_t *crypto_malloc(size_t bytes)
 {
-    return (uint8_t *)malloc(bytes);
+    uint8_t *ptr = (uint8_t *)malloc(bytes);
+
+    if (ptr != nullptr) {
+        memset(ptr, 0xaa, bytes);
+    }
+
+    return ptr;
```
[the remark this answers is not in the document]
Comment (iphydf) [reply toxcore/crypto_core.c]: Removed.

**Review by iphydf — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1877)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @JFreegman, @nurupo, and @sudden6)

___
*[toxcore/crypto_core.c, line 58 at r2](https://reviewable.io/reviews/toktok/c-toxcore/1877#-MtJ9jOIA5qSVz-vEqrA:-MtJF1SQ3G3jxPynZBYK:bp2zt83) ([raw file](https://github.com/toktok/c-toxcore/blob/2b399eba46ed6d405a8aade73768c235149bdf5d/toxcore/crypto_core.c#L58)):*
<details><summary><i>Previously, JFreegman wrote…</i></summary><blockquote>

Why does this function exist? It seems like it would be a good opportunity to use memlock here (for another PR maybe)
</blockquote></details>

Done in this PR, it's a small change.
___
*[toxcore/crypto_core.c, line 155 at r2](https://reviewable.io/reviews/toktok/c-toxcore/1877#-MtJ4Oe_7w0To8kJKe6F:-MtJEeDd8yLFAmfggoJr:b-cs3hea) ([raw file](https://github.com/toktok/c-toxcore/blob/2b399eba46ed6d405a8aade73768c235149bdf5d/toxcore/crypto_core.c#L155)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

why did you change the order and size here? if the order is causing errors it might be a bug
</blockquote></details>

Undone.
___
*[toxcore/crypto_core.c, line 200 at r2](https://reviewable.io/reviews/toktok/c-toxcore/1877#-MtJ8vsxCBHvWx_HhYlE:-MtJEcWhDdIbt3F1PAbU:b-5k7thi) ([raw file](https://github.com/toktok/c-toxcore/blob/2b399eba46ed6d405a8aade73768c235149bdf5d/toxcore/crypto_core.c#L200)):*
<details><summary><i>Previously, JFreegman wrote…</i></summary><blockquote>

This isn't really part of the PR but we need to make sure `length` is > `crypto_box_MACBYTES` before this memcpy. 
</blockquote></details>

Added an assert.

**Review by sudden6 — COMMENTED** [review]
:lgtm_strong: from my side

> *[Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1877#-:-MtJGPFd2i4vrKTwXU86:bybplqg)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @JFreegman and @nurupo)

**Review by nurupo — CHANGES_REQUESTED** [review]
> Reviewed 9 of 15 files at r1, 1 of 2 files at r2, 1 of 2 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1877)* status: 2 change requests, 0 of 1 approvals obtained (waiting on @iphydf and @JFreegman)

___
*[toxcore/crypto_core.c, line 63 at r5](https://reviewable.io/reviews/toktok/c-toxcore/1877#-MtJIqNu1Y3wivYJwlAI:-MtJIqNu1Y3wivYJwlAJ:b-pjp4no) ([raw file](https://github.com/toktok/c-toxcore/blob/b4ce5633628d385bc059617582f8d58bc81e52f3/toxcore/crypto_core.c#L63)):*
> ```C
> 
>     if (ptr != nullptr) {
>         crypto_memlock(ptr, bytes);
> ```

`crypto_memlock` may fail. If it fails, are we allowed to call `crypto_memunlock` on the memory or is that an undefined behavior?

**Review by nurupo — APPROVED** [review]

**Pull request merged**